Allow MemoryMappedFile to work on character device

wfurt opened the pull request:
This change will push two argument checks to the platform part and it will not enforce capacity limits on Unix if the file is a special character device. 

I don't know if this is sufficient to make GPIO and possibly other HW working but it is now at least possible to map /dev/* entries and read and write.

added test for /dev/zero 
```
$ stat /dev/zero
  File: '/dev/zero'
  Size: 0         	Blocks: 0          IO Block: 4096   character special file
Device: 6h/6d	Inode: 8           Links: 1     Device type: 1,5
Access: (0666/crw-rw-rw-)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2020-01-03 13:30:23.368000000 -0800
Modify: 2020-01-03 13:30:23.368000000 -0800
Change: 2020-01-03 13:30:23.368000000 -0800
 Birth: -
```
fixes https://github.com/dotnet/corefx/issues/30686

wfurt committed:
allow MemoryMeppedFile to work on character device

carlossanlop commented on src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs:
```diff
@@ -18,6 +18,39 @@ public partial class MemoryMappedFile
+                    Interop.Sys.FileStatus status;
+                    if (Interop.Sys.FStat(fileStream.SafeFileHandle, out status) == 0)
+                    {
+                        if ((status.Mode & Interop.Sys.FileTypes.S_IFCHR) != 0)
```
Totally optional - if you think it's just as easy to read, this `if` can be simplified:

```cs
isSpecial = (status.Mode & Interop.Sys.FileTypes.S_IFCHR) != 0;
```

wfurt replied:
updated.

carlossanlop commented on src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Windows.cs:
```diff
@@ -24,6 +24,18 @@ public partial class MemoryMappedFile
             SafeFileHandle? fileHandle = fileStream != null ? fileStream.SafeFileHandle : null;
             Interop.Kernel32.SECURITY_ATTRIBUTES secAttrs = GetSecAttrs(inheritability);
 
+            if (fileStream != null)
```
The original issue was reported for Unix special files. Why are these exceptions being added to the Windows file too?

carlossanlop commented on src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.cs:
```diff
@@ -154,12 +154,6 @@ public static MemoryMappedFile CreateFromFile(string path, FileMode mode, string
                 throw new ArgumentException(SR.Argument_EmptyFile);
             }
 
-            if (access == MemoryMappedFileAccess.Read && capacity > fileStream.Length)
```
If you are deferring this check, then you should missed to remove the condition in line 143, you also moved that one to `CreateCore`:
```cs
if (access == MemoryMappedFileAccess.Write)
{
    throw new ArgumentException(SR.Argument_NewMMFWriteAccessNotAllowed, nameof(access));
}
```

carlossanlop commented on src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Windows.cs:
```diff
@@ -24,6 +24,18 @@ public partial class MemoryMappedFile
+                    throw new ArgumentException(SR.Argument_ReadAccessWithLargeCapacity);
+                }
+
+                if (access == MemoryMappedFileAccess.Write)
```
The `CreateNew` method was throwing exception, which you removed. Now it's going to depend on the `fileStream` being not `null`, which for `CreateNew` is never the case.

wfurt replied:
Touching `CreateNew` was never the intention. I miss that because I only added a test for CreateFromFile with a stream. I removed change to the `CreateNew`  method and I added one more test.

carlossanlop commented on src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.cs:
```diff
@@ -293,11 +277,6 @@ public static MemoryMappedFile CreateNew(string? mapName, long capacity, MemoryM
                 throw new ArgumentOutOfRangeException(nameof(access));
             }
 
-            if (access == MemoryMappedFileAccess.Write)
```
What I mentioned above - (line 290 in this file) when we call `CreateCore`, the first argument is `null` (fileStream). This exception will not be thrown anymore with `CreateNew`.

carlossanlop commented on src/libraries/System.IO.MemoryMappedFiles/tests/MemoryMappedFile.CreateFromFile.Tests.cs:
```diff
@@ -868,5 +869,40 @@ public void ReusingNames_Windows(string name)
+            long viewCapacity = 0xFF;
+
+            using (FileStream fs = new FileStream(device, FileMode.Open, FileAccess.ReadWrite, FileShare.ReadWrite))
+            using (MemoryMappedFile memMap = MemoryMappedFile.CreateFromFile(fs, null, viewCapacity, access, HandleInheritability.None, false))
```
Do we have unit tests that throw when `CreateNew` is called with `access = MemoryMappedFileAccess.Write`? Because if we do, they should fail with the current changes (see my comments above).

wfurt replied:
We have some tests for invalid input but if 'mapName' is specific we expect to throw PNSP and that may come before getting to platform code.

carlossanlop reviewed (changes requested):
Left some comments for you to consider.

wfurt committed:
feedback from review

wfurt commented:
I made some updates @carlossanlop and this should be ready for another pass. Touching `CreateNew` was a mistake and it should be fixed now. 
I did more testing around `MemoryMappedFileAccess.Write` and things are a little bit more tricky. While I don't have a concrete use case for a particular device but my intention was to allow whatever OS can do without adding artificial limitations at .NET layer. 

The mmap call will fail on /dev/zero with EACCESS unless fd is opened with `O_RDWR`.
Touching `GetFileAccess()` may be risky.
I also did more testing with other devices like `/dev/null` and the outcome can vary and it can fail with ENOTSUP. When an invalid name is passed in, it is difficult to preserve existing behavior with a deferred check. 

So I abandoned the idea that we can open the right stream and pass proper flags for all devices using the file name.  If needed, the user can create FileStream as needed and pass more options and we will pass it down to OS for final judgment. I would expect the majority of users to use `MemoryMappedFileAccess.ReadWrite`

carlossanlop commented on src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.cs:
```diff
@@ -154,12 +154,6 @@ public static MemoryMappedFile CreateFromFile(string path, FileMode mode, string
                 throw new ArgumentException(SR.Argument_EmptyFile);
             }
 
-            if (access == MemoryMappedFileAccess.Read && capacity > fileStream.Length)
```
Lines 143-146 in this file:
```cs
if (access == MemoryMappedFileAccess.Write)
{
    throw new ArgumentException(SR.Argument_NewMMFWriteAccessNotAllowed, nameof(access));
}
```

Should you remove those lines, since you deferred that exception to `CreateCore`?

wfurt replied:
I tried to explain it in the comment above @carlossanlop. Let me know if that makes. If not, we can chat next week.

carlossanlop replied:
Thanks for the explanation in person, @wfurt. As discussed, the Write-only exception should be reverted to its original state to avoid affecting existing behavior.

carlossanlop commented on src/libraries/System.IO.MemoryMappedFiles/tests/MemoryMappedFile.CreateFromFile.Tests.cs:
```diff
@@ -868,5 +869,85 @@ public void ReusingNames_Windows(string name)
+                using (MemoryMappedFile memMap = MemoryMappedFile.CreateFromFile(device, FileMode.Open, null, viewCapacity, access))
+                using (MemoryMappedViewAccessor view = memMap.CreateViewAccessor(0, viewCapacity, access))
+                {
+                    if (access != MemoryMappedFileAccess.Write)
```
Lines 935-945 is shared code between the two unit tests. Any chance you could put it in a private method and call it from both unit tests?

carlossanlop commented on src/libraries/System.IO.MemoryMappedFiles/tests/MemoryMappedFile.CreateFromFile.Tests.cs:
```diff
@@ -868,5 +869,85 @@ public void ReusingNames_Windows(string name)
+        [PlatformSpecific(TestPlatforms.AnyUnix)]
+        public void OpenCharacterDeviceAsFile(MemoryMappedFileAccess access)
+        {
+            string device = "/dev/zero";
```
Lines 923-929 has shared code. Any chance you could avoid repeating it? Maybe make the device and viewCapacity private members of the class, or wrap the repeated code with a shared private method.

carlossanlop reviewed (changes requested):
I'm back from vacation. Left a couple of additional comments for you to consider.

lpereira commented on src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs:
```diff
@@ -18,6 +18,34 @@ public partial class MemoryMappedFile
+            bool isSpecial = false;
+            if (fileStream != null)
+            {
+                if (fileStream.Length == 0)
```
Getting the `Length` property of a `FileStream` will also call `FStat()`.  `stat()` isn't the fastest syscall in some file systems.

wfurt replied:
good point. I did not realize that. There is no point in saving `fstat()` with `fstat`. 
I reworked the code to always call fstat() if the stream is not null and then use that for both checks instead of using fileStream.Length as we did before.

wfurt committed:
feedback from review

lpereira commented on src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs:
```diff
@@ -18,6 +18,35 @@ public partial class MemoryMappedFile
+                    throw Interop.GetExceptionForIoErrno(errorInfo);
+                }
+
+                isSpecial = (status.Mode & Interop.Sys.FileTypes.S_IFCHR) != 0;
```
Maybe instead of naming `bool isSpecial`, name it `isRegularFile` and test for `S_IFREG` instead for the sake of clarity?

carlossanlop replied:
@lpereira Are those two symbolic names actually opposites? The documentation says:

> S_IFCHR - This is the file type constant of a character-oriented device file.
> S_IFREG - This is the file type constant of a regular file.

This makes me think that is S_IFREG returns false, we can't say for sure if the file is a character-oriented device file, or a block-oriented file, or a directory, or a symbolic link, or a socket or a fifo/pipe (basically, everything that is not a regular file).

Sources:
- https://pubs.opengroup.org/onlinepubs/007908799/xsh/sysstat.h.html
- https://www.gnu.org/software/libc/manual/html_node/Testing-File-Type.html

lpereira replied:
They're not opposites, no, but the name `isSpecial` doesn't tell me much about which kind of the file it is.  I forgot you can memory map sockets, too, so having this pivot on either "is this a regular file" or not isn't a good idea.

Maybe rename it to `isCharDev` or something of the sort, just to be explicit in kind of "special file" we're dealing with?

wfurt replied:
my plan was to use `isRegularFile`  name (as I like it better) but keep an existing check. 
I also originally look at adding support for block devices as well, but I could not find any actual device supporting MemoryMap. That can be fixed later if we get any real use case.

stephentoub commented:
@wfurt, still working on this?

wfurt committed:
feedback from review

wfurt committed:
Merge branch 'master' of https://github.com/dotnet/runtime into cdev_30686

wfurt committed:
roll-back write only access

wfurt commented:
I updated tests, changed naming and rill-back  write-only support as we discussed while back @carlossanlop. This should be ready for another review round.

carlossanlop commented on src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs:
```diff
@@ -18,6 +18,35 @@ public partial class MemoryMappedFile
+                }
+            }
+
             if (mapName != null)
```
Should we pull this verification to the top of this method? There's no point in doing anything if the user tried to pass a mapName.

carlossanlop replied:
If mapName is not null we can throw much earlier, instead of doing all the checks inside `VerifyMemoryMappedFileAccess`. What do you think, @wfurt?

wfurt replied:
It seems like before we check MemoryMappedFileAccess.Write, then capacity > fileStream.Length, then mapName != null and then we may try fileStream.SetLength. It seems like best option would be to preserver that order and perhaps do more VerifyMemoryMappedFileAccess(). 
It seems like the consensus is that it would be OK to reorder exceptions if multiple parameters are invalid. 
If so, we can do what ever we want. But perhaps move that check to VerifyMemoryMappedFileAccess() as well to keep as much of the validation together?

carlossanlop replied:
Let's preserve the order. I don't think it affects much to do the mapName check before or after.

carlossanlop commented on src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs:
```diff
@@ -35,10 +64,12 @@ public partial class MemoryMappedFile
-                // This map is backed by a file.  Make sure the file's size is increased to be
-                // at least as big as the requested capacity of the map.
-                if (fileStream.Length < capacity)
+
+
```
nit: spacing

carlossanlop commented:
I noticed that in MemoryMappedFile.cs, this change removed the code that deletes the file if it didn't use to exist (when the file was created internally via a FileStream). Now if the user calls the `CreateFromFile` overload that takes a `string path`, the file that gets created internally would never get deleted.

I also realized that the newly added behavior difference between Windows and Unix would not be obvious to understand to someone reading this in the future.

If you allow me to provide a suggestion, I think that instead of deferring the code, we should create a private method called `VerifyMemoryMappedFileAccess`, with a Windows version and a Unix version.

The Windows version would contain:
```cs
private void VerifyMemoryMappedFileAccess(MemoryMappedFileAccess access, int capacity, FileStream fileStream, bool existed)
{
    if (access == MemoryMappedFileAccess.Read && capacity > fileStream.Length)
    {
        if (!existed)
        {
            CleanupFile(fileStream, true, path);
        }
        throw new ArgumentException(SR.Argument_ReadAccessWithLargeCapacity);
    }
}
```

And the Unix methods would contain:
```cs

private bool _isRegularFile = true;

private void VerifyMemoryMappedFileAccess(MemoryMappedFileAccess access, int capacity, FileStream fileStream, bool existed)
{
    Interop.Sys.FileStatus status = default;

    if (fileStream != null)
    {
        int result = Interop.Sys.FStat(fileStream.SafeFileHandle, out status);
        if (result != 0)
        {
            Interop.ErrorInfo errorInfo = Interop.Sys.GetLastErrorInfo();
            throw Interop.GetExceptionForIoErrno(errorInfo);
        }

        _isRegularFile = (status.Mode & Interop.Sys.FileTypes.S_IFCHR) == 0;

        if (_isRegularFile)
        {
            if (access == MemoryMappedFileAccess.Read && capacity > status.Size)
            {
                if (!existed)
                {
                    CleanupFile(fileStream, true, path);
                }
                throw new ArgumentException(SR.Argument_ReadAccessWithLargeCapacity);
            }

            if (access == MemoryMappedFileAccess.Write)
            {
                throw new ArgumentException(SR.Argument_NewMMFWriteAccessNotAllowed, nameof(access));
            }
        }
    }
}
```

Having a private field `_isRegularFile` (only available in Unix) would allow you to keep the code you added in `CreateCore`, into which we only access when `fileStream` is not `null`.

```cs
...
if (capacity > status.Size && isRegularFile)
{
    // This map is backed by a file.  Make sure the file's size is increased to be
    // at least as big as the requested capacity of the map for Write* access.
    ...
```

Finally, the new `VerifyMemoryMappedFileAccess` method would be called in 2 places:

MemoryMappedFile.cs:
```cs
public static MemoryMappedFile CreateFromFile(string path, FileMode mode, string? mapName, long capacity,
                                                                        MemoryMappedFileAccess access)
{
    ...
    // Line 157
    VerifyMemoryMappedFileAccess(access, capacity, fileStream, existed);
    ...
}

...

public static MemoryMappedFile CreateFromFile(FileStream fileStream, string? mapName, long capacity,
                                                        MemoryMappedFileAccess access,
                                                        HandleInheritability inheritability, bool leaveOpen)
{
    ...
    // Line 227
    VerifyMemoryMappedFileAccess(access, capacity, fileStream, existed: true);
    ...
}
```

What do you think, @wfurt?

cc @jozkee

Jozkee commented:
> Now if the user calls the CreateFromFile overload that takes a string path, the file that gets created internally would never get deleted.

It would be good to add a test for that.

stephentoub commented on src/libraries/System.IO.MemoryMappedFiles/tests/MemoryMappedFile.CreateFromFile.Tests.cs:
```diff
@@ -868,5 +869,80 @@ public void ReusingNames_Windows(string name)
+        [PlatformSpecific(TestPlatforms.AnyUnix)]
+        public void OpenCharacterDeviceAsStream(MemoryMappedFileAccess access)
+        {
+            string device = "/dev/zero";
```
```suggestion
            const string device = "/dev/zero";
```

stephentoub commented on src/libraries/System.IO.MemoryMappedFiles/tests/MemoryMappedFile.CreateFromFile.Tests.cs:
```diff
@@ -868,5 +869,80 @@ public void ReusingNames_Windows(string name)
+                if (access != MemoryMappedFileAccess.Write)
+                {
+                    byte b = view.ReadByte(0);
+                    // /dev/zero return zeroes.
```
Could we also validate something that doesn't just return 0s, e.g. validate with /dev/urandom that we get back some non-zero values?

wfurt replied:
That is tricky. We cannot write to /dev/(u)random so we would need to split the test. And we can get randomly 0 as well. I can add comment that the test verifies ability to read, not the content it self.

stephentoub commented on src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.cs:
```diff
@@ -154,12 +154,6 @@ public static MemoryMappedFile CreateFromFile(string path, FileMode mode, string
-            {
-                CleanupFile(fileStream, existed, path);
-                throw new ArgumentException(SR.Argument_ReadAccessWithLargeCapacity);
-            }
```
This is changing the order of exceptions we throw, but I think that's ok.  Just wanted to call out that in theory someone could observe it.

carlossanlop replied:
@stephentoub do we need to document exception order alterations as breaking changes?

stephentoub replied:
The same exception still happens if you get something wrong, it's just if you get multiple things wrong you may not get the same error you used to. I think it's probably below the threshold of what we need to document, i.e. no, but I'm open to other opinions... @ericstj, what do you think?

danmosemsft replied:
I believe we've done this a few times before and didn't document it, for what it's worth.

ericstj replied:
In desktop we'd avoid this type of change completely.  It doesn't hurt to document it.

wfurt replied:
I think we can preserve the order so I'll do that to decrease risk.

stephentoub reviewed (commented):
Other than Carlos' suggestion, generally looks good.

wfurt committed:
Merge branch 'master' of https://github.com/dotnet/runtime into cdev_30686

wfurt committed:
feedback from review

wfurt committed:
fix windows

wfurt commented on src/libraries/System.IO.MemoryMappedFiles/tests/MemoryMappedFile.CreateFromFile.Tests.cs:
```diff
@@ -868,5 +869,80 @@ public void ReusingNames_Windows(string name)
+                }
+            }
+            catch (UnauthorizedAccessException) { }
+            // ENODEV Operation not supported by device.
```
The ENODEV is coming from OSX. I did not test on FreeBSD.  I could make the test Platform specific. But since the code is now unix specific the test verifies that at least we do not throw some unexpected exception.

wfurt commented:
I made changes in spirit @carlossanlop suggested and did small updates @stephentoub pointed out. Test seems to be failing for unrelated reasons.

carlossanlop reviewed (approved):
LGTM @wfurt, thanks for making this change. I left one more question in case you want to consider it, but it's not blocking.

Commit:
avoid exception reordering

Commit:
Merge branch 'master' of https://github.com/dotnet/runtime into cdev_30686

wfurt commented:
I look again and there are three places where CreateCore() is called - two with FileStream and one without. I moved the remaining cage to VerifyMemoryMappedFileAccess() so we can preserve order of the checks. Also the in the case when CreateFromFile is called with string, the CreateCore() is already wrapped in try/catch and it has cleanup so I don't think we need it in the VerifyMemoryMappedFileAccess() function.

stephentoub commented on src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Windows.cs:
```diff
@@ -11,12 +11,27 @@ namespace System.IO.MemoryMappedFiles
 {
     public partial class MemoryMappedFile
     {
+
```
```suggestion
```

stephentoub commented on src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Windows.cs:
```diff
@@ -11,12 +11,27 @@ namespace System.IO.MemoryMappedFiles
     public partial class MemoryMappedFile
     {
+
+        // This will verify file access and return file size.
```
It doesn't appear to be returning anything.  Stale comment?

wfurt committed:
feedback from review
